persist/fs/commitlog/reader.go

```diff
@@ -109,58 +165,171 @@ func (r *reader) Read() (
 	annotation ts.Annotation,
 	resultErr error,
 ) {
-	entry, err := r.readEntry()
-	if err != nil {
-		resultErr = err
-		return
+	if r.nextIndex == 0 {
+		go r.readLoop()
+		for i, decoderBuf := range r.decoderBufs {
+			go r.decoderLoop(decoderBuf, r.outBufs[i])
+		}
 	}
+	// Data is written into these channels in round-robin fashion, so reading them
+	// one at a time in the same order results in an ordered stream.
```

Comment:
Not sure I follow entirely. We round-robin across the slice, and write data to each chan. And are subsequently reading decoding each channel's contents. That guarantees we maintain global order iff the readers of each chan are synchronised, I don't see how we're guaranteeing that. 

i.e. say we have concurrency 2, and initial structure -
```
r.outBufs:
  [0]: [] // using an empty slice to indicate channel contents here
  [1]: []
```

after we do a bunch of reads, lets say `a`, `b`, `c`, `d`, we'd end up like:

```
r.outBufs:
  [0]: [a, c] 
  [1]: [b, d]
```

The decoders are only going to maintain local order (i.e. `c` after `a`, and `d` after `b`), not global order. Is that going to cause issues?

Reply:
This works because when we read out of the channels, we also read in round-robin order so with the example you gave

```
r.outBufs:
  [0]: [a, c] 
  [1]: [b, d]
```

We'll read in this order:

1) 0
2) 1
3) 0
4) 1

which translates into a, b, c, d

Reply:
Makes sense. Could you put a little blurb explaining this in the struct/method

Reply:
will do